language/move-vm/types/src/values/values_impl.rs

```diff
@@ -1811,6 +1832,91 @@ impl Vector {
     }
 }
 
+/***************************************************************************************
+ *
+ * Runtime Type checks
+ *
+ **************************************************************************************/
+
+impl Value {
+    pub fn check_type(&self, ty: &Type) -> PartialVMResult<()> {
+        self.0.check_type(ty)
+    }
+}
+
+impl ValueImpl {
+    fn check_type(&self, ty: &Type) -> PartialVMResult<()> {
+        match (self, ty) {
+            (ValueImpl::U8(_), Type::U8)
+            | (ValueImpl::U128(_), Type::U128)
+            | (ValueImpl::U64(_), Type::U64)
+            | (ValueImpl::Bool(_), Type::Bool)
+            | (ValueImpl::Address(_), Type::Address) => Ok(()),
+            (ValueImpl::IndexedRef(r), Type::Reference(ty) | Type::MutableReference(ty)) => {
+                r.copy_value().read_ref()?.check_type(ty)
```

Comment:
This is very concerning... `read_ref` makes a deep copy and can get super expensive if your value is big. Can we find a way to check the value type without calling `read_ref`?